share/wake/lib/system/runner.wake

```diff
@@ -55,97 +56,70 @@ export tuple RunnerOutput =
+export def makeRunner (name: String) (run: Job => RunnerInput => Result RunnerOutput Error): Runner =
+    def do job maybeInput = match maybeInput
+        Fail e ->
+            def _ = primJobFailLaunch job e
 
-            # Don't run any 'post' steps until the Job has stopped running
-            waitJobMerged final job
+            Fail e
+        Pass input -> run job input
 
-    Runner name score doit
+    Runner name do
 
-# This runner does not detect inputs/outputs on it's own
-# You must use Fn{Inputs,Outputs} to fill in this information
+# localRunner: A runner that provides no sandboxing protections and no file tracking.
+#
+# You must use Fn{Inputs,Outputs} to fill in this information for wake to maintain safety and reusability
+# Advanced usage only, proceed with caution
 export def localRunner: Runner =
-    def launch job dir stdin env cmd status runtime cputime membytes ibytes obytes isatty =
-        prim "job_launch"
+    def run (job: Job) ((RunnerInput _ cmd vis env dir stdin _ _ predict isatty): RunnerInput): Result RunnerOutput Error =
+        def jobKey = JobKey dir stdin env.implode cmd.implode 0 "" isatty.booleanToInteger
+        def _ = primJobLaunch job jobKey predict
 
-    def badlaunch job error = prim "job_fail_launch"
+        job
+        | getJobReality
+        |< (\reality RunnerOutput (vis | map getPathName) Nil reality)
 
-    def doit job = match _
-        Fail e ->
-            def _ = badlaunch job e
-
-            Fail e
-        Pass (RunnerInput _ cmd vis env dir stdin _ _ predict isatty) ->
-            def Usage status runtime cputime mem in out = predict
-
-            def _ =
-                launch
-                job
-                dir
-                stdin
-                env.implode
-                cmd.implode
-                status
-                runtime
-                cputime
-                mem
-                in
-                out
-                (booleanToInteger isatty)
-
-            match (getJobReality job)
-                Pass reality -> Pass (RunnerOutput (map getPathName vis) Nil reality)
-                Fail f -> Fail f
-
-    def score _ = Pass 1.0
-
-    Runner "local" score doit
+    makeRunner "local" run
 
+# virtualRunner: A runner that immediatly marks the job as complete using the predicted usage
+#
+# This runner is useful for tracking a unit of work that is job like but not launched as a process
 export def virtualRunner: Runner =
-    def virtual job stdout stderr status runtime cputime membytes ibytes obytes = prim "job_virtual"
-    def badlaunch job error = prim "job_fail_launch"
+    def run (job: Job) ((RunnerInput _ _ vis _ _ _ _ _ predict _): RunnerInput): Result RunnerOutput Error =
+        def _ = primJobVirtual job "" "" predict
 
-    def doit job = match _
-        Fail e ->
-            def _ = badlaunch job e
+        job
+        | getJobReality
+        |< (\reality RunnerOutput (vis | map getPathName) Nil reality)
```

Comment:
Very minor, but given my functional instincts I'd be inclined to eta(?) reduce this.
```suggestion
        job.getJobReality
        |< RunnerOutput (vis | map getPathName) Nil
```